Extensions/JPCFunction/JPCFunction.m

```diff
@@ -50,19 +57,112 @@ + (void)defineCFunction:(NSString *)funcName types:(NSString *)types
+        JP_FFI_RETURN_CASE('L', unsigned long, numberWithUnsignedLong)
+        JP_FFI_RETURN_CASE('q', long long, numberWithLongLong)
+        JP_FFI_RETURN_CASE('Q', unsigned long long, numberWithUnsignedLongLong)
+        JP_FFI_RETURN_CASE('f', float, numberWithFloat)
+        JP_FFI_RETURN_CASE('F', CGFloat, numberWithCGFloat)
+        JP_FFI_RETURN_CASE('d', double, numberWithDouble)
+        JP_FFI_RETURN_CASE('B', BOOL, numberWithBool)
+        case '^': {
+            JPBoxing *box = [[JPBoxing alloc] init];
+            box.pointer = (*(void**)src);
+            return box;
+        }
+        case '@':
+        case '#': {
+            return (__bridge id)(*(void**)src);
+        }
+        case '{': {
+            NSString *structName = [NSString stringWithCString:typeString encoding:NSASCIIStringEncoding];
+            structName = [structName substringWithRange:NSMakeRange(1, structName.length - 2)];
+            NSDictionary *structDefine = [JPExtension registeredStruct][structName];
+            id dict = [JPExtension getDictOfStruct:src structDefine:structDefine];
+            id ret = [JSValue valueWithObject:dict inContext:[JPEngine context]];
+            return ret;
+        }
+        default:
+            return nil;
+    }
+}
+
++ (void)convertObject:(id)object toCValue:(void *)dist forType:(const char *)typeString
+{
+#define JP_CALL_ARG_CASE(_typeString, _type, _selector)\
+    case _typeString:{\
+        *(_type *)dist = [(NSNumber *)object _selector];\
+        break;\
+    }
+    switch (typeString[0]) {
+        JP_CALL_ARG_CASE('c', char, charValue)
+        JP_CALL_ARG_CASE('C', unsigned char, unsignedCharValue)
+        JP_CALL_ARG_CASE('s', short, shortValue)
+        JP_CALL_ARG_CASE('S', unsigned short, unsignedShortValue)
+        JP_CALL_ARG_CASE('i', int, intValue)
+        JP_CALL_ARG_CASE('I', unsigned int, unsignedIntValue)
+        JP_CALL_ARG_CASE('l', long, longValue)
+        JP_CALL_ARG_CASE('L', unsigned long, unsignedLongValue)
+        JP_CALL_ARG_CASE('q', long long, longLongValue)
+        JP_CALL_ARG_CASE('Q', unsigned long long, unsignedLongLongValue)
+        JP_CALL_ARG_CASE('f', float, floatValue)
+        JP_CALL_ARG_CASE('F', CGFloat, CGFloatValue)
+        JP_CALL_ARG_CASE('d', double, doubleValue)
+        JP_CALL_ARG_CASE('B', BOOL, boolValue)
+        case '^': {
+            void *ptr = [((JPBoxing *)object) unboxPointer];
+            *(void **)dist = ptr;
+            break;
+        }
+        case '#':
+        case '@': {
+            id ptr = object;
+            *(void **)dist = (__bridge void *)(ptr);
+            break;
+        }
+        case '{': {
+            NSString *structName = [NSString stringWithCString:typeString encoding:NSASCIIStringEncoding];
+            structName = [structName substringWithRange:NSMakeRange(1, structName.length - 2)];
+            NSDictionary *structDefine = [JPExtension registeredStruct][structName];
```

Comment:
Seems use some trick to map a native struct signature to js struct sigature. But maybe  better to compare and alter some error when these two  signature not match.